From 03c98fc878f042ab0eb8cb9b6f222ce2c9128a1a Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Mon, 6 Nov 2023 15:50:58 +0100
Subject: [PATCH 0049/1695] core: dt_driver: reference output device reference
 as void *

Changes dt_driver API function to reference device reference as
void * instead of void ** which could be confusing as the reference
can be a pointer to a device pointer (e.g. in clk_dt.c) or a pointer
to a structure (e.g. interrupt.c).

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 6dd291e2cea1bfd79d6de884a7ca2086e6a490b1 core: dt_driver: reference output device reference as void *]
Change-Id: I1ed518553b8dadf01bde5bdca4d4da06c7bc6ea7
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
[etienne: cherry-picked from b548a65701d85e9c4fdf7a08f56afdf93ddacf5e]
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/include/kernel/dt_driver.h | 14 +++++++-------
 core/kernel/dt_driver.c         | 19 +++++++++++--------
 2 files changed, 18 insertions(+), 15 deletions(-)

diff --git a/core/include/kernel/dt_driver.h b/core/include/kernel/dt_driver.h
index d95f8b94c..0eef55db0 100644
--- a/core/include/kernel/dt_driver.h
+++ b/core/include/kernel/dt_driver.h
@@ -102,7 +102,7 @@ struct dt_pargs {
  *
  * @parg: phandle argument(s) referencing the device in the FDT.
  * @data: driver private data registered in struct dt_driver.
- * @out_device: output device reference upon success, e.g. a struct clk
+ * @device_ref: output device reference upon success, e.g. a struct clk
  *	pointer for a clock driver.
  *
  * Return code:
@@ -111,7 +111,7 @@ struct dt_pargs {
  * Any TEE_Result compliant code in case of error.
  */
 typedef TEE_Result (*get_of_device_func)(struct dt_pargs *parg, void *data,
-					 void **out_device);
+					 void *device_ref);
 
 /**
  * dt_driver_register_provider - Register a driver provider
@@ -139,7 +139,7 @@ TEE_Result dt_driver_register_provider(const void *fdt, int nodeoffset,
  * @nodeoffset: node offset in the FDT
  * @prop_idx: index of the phandle data in the property
  * @type: Driver type
- * @out_device: output device opaque reference upon support, for example
+ * @device_ref: output device opaque reference upon support, for example
  *	a struct clk pointer for a clock driver.
 
  * Return code:
@@ -153,7 +153,7 @@ TEE_Result dt_driver_device_from_node_idx_prop(const char *prop_name,
 					       const void *fdt, int nodeoffset,
 					       unsigned int prop_idx,
 					       enum dt_driver_type type,
-					       void **out_device);
+					       void *device_ref);
 
 /*
  * dt_driver_device_from_parent - Return a device instance based on the parent.
@@ -163,7 +163,7 @@ TEE_Result dt_driver_device_from_node_idx_prop(const char *prop_name,
  * @fdt: FDT base address
  * @nodeoffset: node offset in the FDT
  * @type: Driver type
- * @out_device: output device opaque reference upon success, for example
+ * @device_ref: output device opaque reference upon success, for example
  *	a struct i2c_dev pointer for a I2C bus driver
  *
  * Return code:
@@ -173,7 +173,7 @@ TEE_Result dt_driver_device_from_node_idx_prop(const char *prop_name,
  */
 TEE_Result dt_driver_device_from_parent(const void *fdt, int nodeoffset,
 					enum dt_driver_type type,
-					void **out_device);
+					void *device_ref);
 
 /*
  * dt_driver_device_from_node_idx_prop_phandle() - Same as
@@ -190,7 +190,7 @@ TEE_Result dt_driver_device_from_node_idx_prop_phandle(const char *prop_name,
 						       unsigned int prop_index,
 						       enum dt_driver_type type,
 						       uint32_t phandle,
-						       void **out_device);
+						       void *device_ref);
 
 /*
  * dt_driver_get_crypto() - Request crypto support for driver initialization
diff --git a/core/kernel/dt_driver.c b/core/kernel/dt_driver.c
index 59201f61a..4ffac3785 100644
--- a/core/kernel/dt_driver.c
+++ b/core/kernel/dt_driver.c
@@ -249,7 +249,8 @@ dt_driver_get_provider_by_phandle(uint32_t phandle, enum dt_driver_type type)
 
 static TEE_Result device_from_provider_prop(struct dt_driver_provider *prv,
 					    const void *fdt, int phandle_node,
-					    const uint32_t *prop, void **device)
+					    const uint32_t *prop,
+					    void *device_ref)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
 	struct dt_pargs *pargs = NULL;
@@ -266,7 +267,7 @@ static TEE_Result device_from_provider_prop(struct dt_driver_provider *prv,
 	for (n = 0; n < prv->provider_cells; n++)
 		pargs->args[n] = fdt32_to_cpu(prop[n]);
 
-	res = prv->get_of_device(pargs, prv->priv_data, device);
+	res = prv->get_of_device(pargs, prv->priv_data, device_ref);
 
 	free(pargs);
 
@@ -274,7 +275,8 @@ static TEE_Result device_from_provider_prop(struct dt_driver_provider *prv,
 }
 
 TEE_Result dt_driver_device_from_parent(const void *fdt, int nodeoffset,
-					enum dt_driver_type type, void **device)
+					enum dt_driver_type type,
+					void *device_ref)
 {
 	int parent = -1;
 	struct dt_driver_provider *prv = NULL;
@@ -291,7 +293,8 @@ TEE_Result dt_driver_device_from_parent(const void *fdt, int nodeoffset,
 		return TEE_ERROR_DEFER_DRIVER_INIT;
 	}
 
-	return device_from_provider_prop(prv, fdt, nodeoffset, NULL, device);
+	return device_from_provider_prop(prv, fdt, nodeoffset, NULL,
+					 device_ref);
 }
 
 TEE_Result dt_driver_device_from_node_idx_prop_phandle(const char *prop_name,
@@ -300,7 +303,7 @@ TEE_Result dt_driver_device_from_node_idx_prop_phandle(const char *prop_name,
 						       unsigned int prop_index,
 						       enum dt_driver_type type,
 						       uint32_t phandle,
-						       void **device)
+						       void *device_ref)
 {
 	int len = 0;
 	const uint32_t *prop = NULL;
@@ -328,14 +331,14 @@ TEE_Result dt_driver_device_from_node_idx_prop_phandle(const char *prop_name,
 		return TEE_ERROR_ITEM_NOT_FOUND;
 
 	return device_from_provider_prop(prv, fdt, phandle_node_unused,
-					 prop + prop_index, device);
+					 prop + prop_index, device_ref);
 }
 
 TEE_Result dt_driver_device_from_node_idx_prop(const char *prop_name,
 					       const void *fdt, int nodeoffset,
 					       unsigned int prop_idx,
 					       enum dt_driver_type type,
-					       void **device)
+					       void *device_ref)
 {
 	int len = 0;
 	int idx = 0;
@@ -399,7 +402,7 @@ TEE_Result dt_driver_device_from_node_idx_prop(const char *prop_name,
 		idx32++;
 
 		return device_from_provider_prop(prv, fdt, phandle_node,
-						 prop + idx32, device);
+						 prop + idx32, device_ref);
 	}
 
 	return TEE_ERROR_ITEM_NOT_FOUND;
-- 
2.47.2

